lib/SIL/IR/OperandOwnership.cpp

```diff
@@ -399,20 +408,26 @@ OperandOwnershipClassifier::visitStoreBorrowInst(StoreBorrowInst *i) {
   return OperandOwnership::TrivialUse;
 }
 
-static OperandOwnership getFunctionArgOwnership(SILArgumentConvention argConv) {
+OperandOwnership OperandOwnershipClassifier::getFunctionArgOwnership(
+  SILArgumentConvention argConv,
+  bool hasCallerScope) {
+
   switch (argConv) {
   case SILArgumentConvention::Indirect_In:
   case SILArgumentConvention::Direct_Owned:
     return OperandOwnership::ForwardingConsume;
 
-  // A guaranteed argument is forwarded into the callee. However, from the
+  // A guaranteed argument is forwarded into the callee. If the call itself has
+  // no scope (i.e. it is a single apply, try_apply or yield), then from the
   // caller's point of view it looks like an instantaneous use. Consequently,
   // owned values may be passed to guaranteed arguments without an explicit
-  // borrow scope in the caller.
+  // borrow scope in the caller. A begin_apply, on the other hand, does have an
+  // explicit borrow scope in the caller.
```

Comment:
Can you finish the sentence in this comment? My thought:

```
In contrast, a begin_apply /does/ have an explicit borrow scope in the caller so we must treat arguments passed to it as being borrowed for the entire borrow scope.
```

Also what about renaming  hasCallerScope to hasScopeInCaller? I think it may make it clear that you are not talking about something called a "caller scope" but more specifically something called a scope that is in the caller. But if you are against that I am ok with it.

Reply:
On another note, I was also thinking about this specific commit and I realized that someone reading it might be thinking that this might have been fixing a miscompile when I think that this is really just making the info more descriptive (the ownership verifier uses BorrowedOperand to discover the end of lifetime uses for borrowed operands, so we were always verifying this correctly). If you agree with my worry, can you add to this specific commit message that this is only enriching the info. Your thoughts?

Reply:
I applied your suggestions verbatim. These changes to operand ownership all fix bugs when we canonicalize OSSA. Previously we were only using it for verification though, so didn't need to be perfect for correctness.

Reply:
I'll fix the commit message.